### PR TITLE
feat: restrict non credential provider interactions

### LINKED ISSUE #712: feat: restrict non credentials interactions
### Describe the feature you'd like to request

The following things should be restricted to credentials only accounts:

- [x] Invites should only be available if credentials login enabled
- [x] Any manage functionallity on users (except language selection) should be disabled for non credentials accounts
- [x] Users of non credentials provider can not be added / removed from groups manually
- [x] Hide pages like `Security` page in user preferences for non credentials users

### Describe the solution you'd like to see

- [x] Add field `provider` to user in database
- [x] Implement above mentioned restrictions
- [x] Add alerts to inform user on why certain fields can not be modified (user preferences) / certain users membership on groups can not be managed.

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by Meierschlumpf**
<br/>
<div align="center">
  
  <h3>Homarr</h3>
</div>

**Thank you for your contribution. Please ensure that your pull request meets the following pull request:**

- [x] Builds without warnings or errors (``pnpm buid``, autofix with ``pnpm format:fix``)
- [x] Pull request targets ``dev`` branch
- [x] Commits follow the [conventional commits guideline](https://www.conventionalcommits.org/en/v1.0.0/)
- [x] No shorthand variable names are used (eg. ``x``, ``y``, ``i`` or any abbrevation)

Closes #712

image

image

### CODE REVIEW

File: apps/nextjs/src/app/[locale]/manage/users/[userId]/general/page.tsx [hunk 1]
```diff
@@ -53,27 +54,37 @@ export default async function EditUserPage({ params }: Props) {
...
-        </Box>
-        <Box w={{ base: "100%", lg: 200 }}>
-          <UserProfileAvatarForm user={user} />
-        </Box>
-      </Group>
+      <Alert variant="light" color="yellow" icon={<IconExclamationCircle size="1rem" stroke={1.5} />}>
+        {t("management.page.user.fieldsDisabledExternalProvider")}
+      </Alert>
+
+      <Stack>
```
Comment (Meierschlumpf) [inline apps/nextjs/src/app/[locale]/manage/users/[userId]/general/page.tsx:65]: Inner stack can be removed

File: packages/api/src/router/user.ts [hunk 2]
```diff
@@ -7,9 +7,12 @@ import { groupMembers, groupPermissions, groups, invites, users } from "@homarr/
 import { validation, z } from "@homarr/validation";
 
 import { createTRPCRouter, protectedProcedure, publicProcedure } from "../trpc";
+import { assertCredentialsEnabled } from "./invite/asserts";
 
 export const userRouter = createTRPCRouter({
   initUser: publicProcedure.input(validation.user.init).mutation(async ({ ctx, input }) => {
+    assertCredentialsEnabled();
```
Comment (manuel-rw) [inline packages/api/src/router/user.ts:14]: assert is a name that should only be used for test.
I suggest ``throwIfCredentialsDisabled()``

File: packages/api/src/router/user.ts [hunk 3]
```diff
@@ -190,26 +212,34 @@ export const userRouter = createTRPCRouter({
...
-      }
+    if (dbUser.provider !== "credentials") {
+      throw new TRPCError({
+        code: "FORBIDDEN",
+        message: "Password can not be changed for users with external providers",
+      });
+    }
 
+    // Admins can change the password of other users without providing the previous password
+    const isPreviousPasswordRequired = ctx.session.user.id === input.userId;
```
Comment (manuel-rw) [inline packages/api/src/router/user.ts:240]: Can you log an information whether this was required or not?

**Review by manuel-rw — CHANGES_REQUESTED** [review]

File: packages/api/src/router/user.ts (hunk 2 above)
Comment (Meierschlumpf) [reply packages/api/src/router/user.ts:15]: I don't full agree. In my opinion it should be placed whereever we want to restrict the continous execution to something that's also why typescript code like this exists:

```ts
const assertNotNullable = <T>(input: T): asserts input is NonNullable<T> => {
  if (input === null || input === undefined) {
    throw new Error("Value is null or undefined");
  }
}
```

but as I don't use this assertion I'll rename it

Comment (manuel-rw) [reply packages/api/src/router/user.ts]: Okay, I see your point.
But regardless, ``throwIfXXXX()`` seems more descriptive than ``assertXXX`` in my opinion.
Up to you what name you want to keep

**Review by manuel-rw — APPROVED** [review]

**Pull request merged**